File: packages/arcis-python/arcis/middleware/signup_protection.py
"""
Arcis signup-form protection.

Composite primitive combining email validation (syntax + disposable),
bot detection, and a dedicated per-IP rate limit. Matches the Arcjet
`protectSignup` convenience API while staying fully local (no cloud
lookups).

Example:
    protection = SignupProtection(rate_limit_max=5, rate_limit_window_ms=60_000)
    result = protection.check(request)
    if not result.allowed:
        raise HTTPException(status_code=400, detail=result.reason)
"""

from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional

from ..validation.email import validate_email_address
from .bot_detection import detect_bot
from .rate_limit import RateLimiter, RateLimitExceeded


SignupBlockReason = str  # 'missing_email' | 'invalid_email' | 'disposable_email' | 'bot' | 'rate_limited' | 'ok'


@dataclass
class SignupCheckResult:
    """Outcome of a signup protection check."""
    allowed: bool
    reason: SignupBlockReason
    details: Dict[str, Any] = field(default_factory=dict)


def check_signup(
    request: Any,
    *,
    email_field: str = "email",
    check_email: bool = True,
    block_disposable: bool = True,
    check_bot: bool = True,
    allowed_bot_categories: Optional[List[str]] = None,
    allowed_email_domains: Optional[List[str]] = None,
    blocked_email_domains: Optional[List[str]] = None,
) -> SignupCheckResult:
    """Pure signup check. No rate-limit mutation; safe to call repeatedly.

    `request` must expose either a dict-like `body`/`json`/`form` attribute
    or be a mapping itself. The email field is looked up in that order.
    """
    allowed_bot_categories = allowed_bot_categories or []

    if check_bot:
        bot = detect_bot(request)
        if bot.is_bot and bot.category not in allowed_bot_categories:
            return SignupCheckResult(
                allowed=False,
                reason="bot",
                details={"category": bot.category, "name": bot.name, "confidence": bot.confidence},
            )

    if check_email:
        email = _extract_field(request, email_field)
        if not isinstance(email, str) or not email:
            return SignupCheckResult(allowed=False, reason="missing_email")

        result = validate_email_address(
            email,
            check_disposable=block_disposable,
            allowed_domains=allowed_email_domains,
            blocked_domains=blocked_email_domains,
        )
        if not result.valid:
            reason = "disposable_email" if result.reason == "disposable" else "invalid_email"
            return SignupCheckResult(
                allowed=False,
                reason=reason,
                details={"email_reason": result.reason},
            )

    return SignupCheckResult(allowed=True, reason="ok")


def _extract_field(request: Any, field_name: str) -> Any:
    """Best-effort extraction of a body field across request shapes."""
    for attr in ("body", "json", "form"):
        container = getattr(request, attr, None)
        if callable(container):
            try:
                container = container()
            except Exception:
                container = None
        if isinstance(container, dict) and field_name in container:
            return container[field_name]
    if isinstance(request, dict) and field_name in request:
        return request[field_name]
    return None


class SignupProtection:
    """Stateful signup protection: bundles a rate limiter with the pure check.

    Call `.check(request)` per request; on exhaustion it returns a
    `SignupCheckResult` with reason='rate_limited' rather than raising.

    Remember to call `.close()` on shutdown so the rate-limiter cleanup
    thread exits cleanly.
    """

    def __init__(
        self,
        *,
        email_field: str = "email",
        check_email: bool = True,
        block_disposable: bool = True,
        check_bot: bool = True,
        allowed_bot_categories: Optional[List[str]] = None,
        allowed_email_domains: Optional[List[str]] = None,
        blocked_email_domains: Optional[List[str]] = None,
        rate_limit_max: Optional[int] = 5,
        rate_limit_window_ms: int = 60_000,
        rate_limit_key_func: Optional[Callable[[Any], str]] = None,
        on_blocked: Optional[Callable[[Any, SignupCheckResult], None]] = None,
    ):
        self._opts = dict(
            email_field=email_field,
            check_email=check_email,
            block_disposable=block_disposable,
            check_bot=check_bot,
            allowed_bot_categories=allowed_bot_categories,
            allowed_email_domains=allowed_email_domains,
            blocked_email_domains=blocked_email_domains,
        )
        self._on_blocked = on_blocked
        self._limiter: Optional[RateLimiter] = None
        if rate_limit_max is not None and rate_limit_max > 0:
            self._limiter = RateLimiter(
                max_requests=rate_limit_max,
                window_ms=rate_limit_window_ms,
                message="Too many signup attempts",
                key_func=rate_limit_key_func,
            )

    def check(self, request: Any) -> SignupCheckResult:
        result = check_signup(request, **self._opts)
        if not result.allowed:
            if self._on_blocked:
                self._on_blocked(request, result)
            return result

        if self._limiter is not None:
            try:
                self._limiter.check(request)
            except RateLimitExceeded as e:
                rl_result = SignupCheckResult(
                    allowed=False,
                    reason="rate_limited",
                    details={"retry_after": getattr(e, "retry_after", None)},
                )
                if self._on_blocked:
                    self._on_blocked(request, rl_result)
                return rl_result

        return result

    def close(self) -> None:
        if self._limiter is not None:
            self._limiter.close()

File: packages/arcis-node/src/middleware/signup-protection.ts
/**
 * @module @arcis/node/middleware/signup-protection
 *
 * Composite signup-form protection: one middleware that combines email
 * validation (syntax + disposable), bot detection, and a dedicated
 * per-IP rate limit. Matches the Arcjet `protectSignup` convenience
 * primitive but stays fully local — no cloud lookups.
 *
 * @example
 *   app.post('/signup', signupProtection(), handler);
 *
 * @example
 *   app.post('/signup', signupProtection({
 *     emailField: 'email',
 *     rateLimit: { max: 5, windowMs: 60_000 },
 *     blockDisposable: true,
 *   }), handler);
 */

import type { Request, Response, NextFunction, RequestHandler } from 'express';
import { validateEmail } from '../validation/email';
import { detectBot, type BotCategory } from './bot-detection';
import { createRateLimiter } from './rate-limit';

export type SignupBlockReason =
  | 'missing_email'
  | 'invalid_email'
  | 'disposable_email'
  | 'bot'
  | 'rate_limited';

export interface SignupCheckResult {
  allowed: boolean;
  reason: SignupBlockReason | 'ok';
  details?: Record<string, unknown>;
}

export interface SignupProtectionOptions {
  /** Request body field holding the email address. Default: 'email' */
  emailField?: string;
  /** Run email validation. Default: true */
  checkEmail?: boolean;
  /** Reject disposable email domains. Default: true */
  blockDisposable?: boolean;
  /** Run bot detection. Default: true */
  checkBot?: boolean;
  /** Bot categories allowed through (e.g. test harnesses). Default: [] — all bots blocked */
  allowedBotCategories?: BotCategory[];
  /** Per-IP rate limit on signup endpoint. Set to `false` to disable. Default: 5 requests / 60s */
  rateLimit?: { max?: number; windowMs?: number } | false;
  /** Extra email domains to allow (bypasses disposable check) */
  allowedEmailDomains?: string[];
  /** Extra email domains to block */
  blockedEmailDomains?: string[];
  /** Called when a request is blocked — for telemetry/logging */
  onBlocked?: (req: Request, result: SignupCheckResult) => void;
}

export interface SignupProtectionMiddleware extends RequestHandler {
  /** Release the rate-limiter cleanup interval */
  close: () => void;
}

/**
 * Pure signup check — no rate-limit mutation, no response writes.
 * Useful for framework adapters or custom control flow.
 */
export function checkSignup(
  req: Request,
  options: SignupProtectionOptions = {}
): SignupCheckResult {
  const {
    emailField = 'email',
    checkEmail = true,
    blockDisposable = true,
    checkBot = true,
    allowedBotCategories = [],
    allowedEmailDomains = [],
    blockedEmailDomains = [],
  } = options;

  if (checkBot) {
    const bot = detectBot(req);
    if (bot.isBot && !allowedBotCategories.includes(bot.category)) {
      return {
        allowed: false,
        reason: 'bot',
        details: { category: bot.category, name: bot.name, confidence: bot.confidence },
      };
    }
  }

  if (checkEmail) {
    const email = (req.body as Record<string, unknown> | undefined)?.[emailField];
    if (typeof email !== 'string' || email.length === 0) {
      return { allowed: false, reason: 'missing_email' };
    }
    const result = validateEmail(email, {
      checkDisposable: blockDisposable,
      allowedDomains: allowedEmailDomains,
      blockedDomains: blockedEmailDomains,
    });
    if (!result.valid) {
      const reason: SignupBlockReason =
        result.reason === 'disposable' ? 'disposable_email' : 'invalid_email';
      return { allowed: false, reason, details: { emailReason: result.reason } };
    }
  }

  return { allowed: true, reason: 'ok' };
}

/**
 * Express middleware: applies bot + email + rate-limit checks to a signup
 * endpoint. Responds 400/403/429 with a JSON body on block; otherwise
 * calls `next()`.
 */
export function signupProtection(
  options: SignupProtectionOptions = {}
): SignupProtectionMiddleware {
  const rateLimitCfg = options.rateLimit;
  const limiter =
    rateLimitCfg === false
      ? null
      : createRateLimiter({
          max: rateLimitCfg?.max ?? 5,
          windowMs: rateLimitCfg?.windowMs ?? 60_000,
          message: 'Too many signup attempts',
        });

  const handler: RequestHandler = (req, res, next) => {
    const result = checkSignup(req, options);
    if (!result.allowed) {
      options.onBlocked?.(req, result);
      const status = result.reason === 'bot' ? 403 : 400;
      res.status(status).json({ error: 'signup_blocked', reason: result.reason });
      return;
    }
    if (limiter) {
      // Delegate to rate limiter — it will respond 429 on breach or call next() otherwise.
      let rateLimited = false;
      const rateLimitNext: NextFunction = (err?: unknown) => {
        if (err) return next(err);
        if (!rateLimited) next();
      };
      const patchedRes = new Proxy(res, {
        get(target, prop) {
          if (prop === 'status') {
            return (code: number) => {
              if (code === 429) {
                rateLimited = true;
                options.onBlocked?.(req, { allowed: false, reason: 'rate_limited' });
              }
              return (target.status as (c: number) => Response).call(target, code);
            };
          }
          return Reflect.get(target, prop);
        },
      });
      limiter(req, patchedRes as Response, rateLimitNext);
      return;
    }
    next();
  };

  const middleware = handler as SignupProtectionMiddleware;
  middleware.close = () => limiter?.close();
  return middleware;
}
